src/Agent.Worker/ResourceMetricsManager.cs

```diff
@@ -281,69 +259,29 @@ private async Task GetMemoryInfoAsync()
             }
 
             if (PlatformUtil.RunningOnLinux)
             {
-                // Some compact Linux distributions like UBI may not have "free" utility installed, or it may have a custom output
-                // We don't want to break currently existing pipelines with ADO warnings
-                // so related errors thrown here will be sent to the trace or debug logs by caller methods
-
-                using var processInvoker = HostContext.CreateService<IProcessInvoker>();
-
-                using var timeoutTokenSource = new CancellationTokenSource();
-                timeoutTokenSource.CancelAfter(TimeSpan.FromMilliseconds(METRICS_UPDATE_INTERVAL));
-
-                using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(
-                    _context.CancellationToken,
-                    timeoutTokenSource.Token);
-
-                processInvoker.OutputDataReceived += delegate (object sender, ProcessDataReceivedEventArgs message)
+                string[] memoryInfo = await File.ReadAllLinesAsync("/proc/meminfo", cancellationToken);
+                if (cancellationToken.IsCancellationRequested)
                 {
-                    if (!message.Data.StartsWith("Mem:"))
-                    {
-                        return;
-                    }
-
-                    var processInvokerOutputString = message.Data;
-                    var memoryInfoString = processInvokerOutputString.Split(" ", StringSplitOptions.RemoveEmptyEntries);
-
-                    if (memoryInfoString.Length != 7)
-                    {
-                        throw new Exception("\"free\" utility has non-default output");
-                    }
-
-                    lock (_memoryInfoLock)
-                    {
-                        _memoryInfo.Updated = DateTime.Now;
-                        _memoryInfo.TotalMemoryMB = long.Parse(memoryInfoString[1]);
-                        _memoryInfo.UsedMemoryMB = long.Parse(memoryInfoString[2]);
-                    }
-                };
+                    return;
+                }
 
-                processInvoker.ErrorDataReceived += delegate (object sender, ProcessDataReceivedEventArgs message)
-                {
-                    Trace.Error($"Error on receiving memory info: {message.Data}");
-                };
+                // The /proc/meminfo file contains several memory counters. To calculate the available memory
+                // we need to get the total memory and the available memory counters
+                // The available memory contains the sum of free, cached, and buffer memory
+                // it shows more accurate information about the memory usage than the free memory counter
+                int totalMemory = int.Parse(memoryInfo[0].Split(" ", StringSplitOptions.RemoveEmptyEntries)[1]);
+                int availableMemory = int.Parse(memoryInfo[2].Split(" ", StringSplitOptions.RemoveEmptyEntries)[1]);
```

Comment:
need to check array length, there is no guarantee [0] or [2] exists

Reply:
The reason we don't check the array length here is because we parse `/proc/meminfo`, and the indexes here may not exist only in two cases: 1. if the `/proc/meminfo` is unavailable - which is guarantee that the system itself is broken and most of the services will not work anyway, so the agent fails even before the monitor runs. 2. The output format in `/proc/meminfo` has changed - which happens rarely once in a decade and previously with the updates only the new fields were added, and already existing fields were never removed.

Reply:
There is an exception wrapper on calls to GetCpuInfoAsync, so I'm okay with this.